Added missing funcitonality for statistic parser.

nolyn opened the pull request:
Can now display values in zoom box:
- Values of affine motion vectors
- Motion vectors for geo partitioning mode

nolyn committed:
Addes missing funcitonality for statistic parser. Can now display values in zoom box:
- Values of affine motion vectors
- Motion vectors for geo partitioning mode

nolyn commented:
This is strange, I'm not able to reproduce the build problems. Tried on my system with QtCreator and Vscode. Also on a ubuntu vm, which should be similar to the one used by the runner

nolyn commented:
I also ran into a strange problem that made YUView crash. My settings file was corrupted. I guess it was from a different YUView version. Removing it fixed the crashing

nolyn committed:
addec include for QPolygon

nolyn commented:
ok, so explicitly adding the include for QPolygon produces
```
fatal error: QPolygon: No such file or directory
 #include <QPolygon>
```
in the runner. all works fine locally. what to make of it ...

nolyn committed:
change QPolygon include

nolyn committed:
removed QPolygon include from StatisticsData.h

nolyn committed:
test if adding gui to csv stat test solve compile problem

nolyn committed:
implement containsPoint, similar to the one of QPolygon, but does not
depend on it. Use QVector(QPoint) which is part of QtCore

ChristianFeldmann commented on YUViewLib/src/statistics/StatisticsData.cpp:
```diff
@@ -35,6 +35,7 @@
 #include "common/functions.h"
 
 #include <QPainterPath>
+#include <QtGui/QPolygon>
```
Please no Qt classes in the statisticsData. The ultimate goal is to have this part free of Qt so it can be reused in other projects. (The existing QPainterPath include can be removed too. This is just a remnant of the past).

nolyn replied:
I removed them. Basically took the corresponding code from qt and removed/change the types to std types. Is this a problem with licensing? Think not since its open source and we already mention that we use it.

nolyn replied:
There is still some QPoint, QStringList, QRect. I can remove those as well.

ChristianFeldmann replied:
Hmm you are raising a good point actually. Qt is LGPL I think and our license is GPL. Not sure if this is compatible/lega.

nolyn replied:
I found a source on stackoverflow (see comment in code) which I used instead. I also had to optmized that quite a bit, so now it should be different enough

ChristianFeldmann commented on YUViewLib/src/statistics/StatisticsData.cpp:
```diff
@@ -165,6 +225,60 @@ QStringPairList StatisticsData::getValuesAt(const QPoint &pos) const
+          foundStats = true;
+        }
+      }
+    }
```
We should probably also add a test for this. Do you have some stats file that contains this data?

nolyn replied:
Yes, here are some.  Skateboard_QP22_grepMV.vtmbmsstats contains among others AffineMVL0/1. Skateboard_QP22_grepGeo.vtmbmsstats has some geo statistics that use the polygons. There are some gaps in the statistics, since they are taken from my work on geometry padding (which is not coded, thus no statistics). The statistics themselves are the standard output by VTM though.

[Skateboard_QP22.zip](https://github.com/IENT/YUView/files/6334150/Skateboard_QP22.zip)

ChristianFeldmann reviewed (changes requested):
Aah I think I get it. So this is just about reporting the values for the GEO partitioning polygons when the zoom box is enabled. So the rendering already worked before. Got it. 
And what exactly do you need to calculate here? If a point is within a polygon, right? Just a side qustion: A polygon here can only have 4 corners max, right?

nolyn committed:
removed usage of QVector from StatisticsData

nolyn committed:
small cleanup

nolyn committed:
removed dependence on qt code for check if point is inside polygon

nolyn committed:
fixed corner case: when point had same y as polygon point, was not detected as inside

nolyn commented:
I also had a look at removing the other qt dependencies. So far I removed them for for 
```
YUViewLib/src/statistics/StatisticsData.cpp 
YUViewLib/src/statistics/StatisticsData.h
```
except for the reference to QDebug. Don't know what to replace it with. Also, were to stop? I.e. which part should form the qt indpendent statistics. So far this required changes also in
```
YUViewLib/src/common/typedef.h
YUViewLib/src/playlistitem/playlistItemCompressedVideo.cpp
YUViewLib/src/playlistitem/playlistItemStatisticsFile.cpp
YUViewLib/src/statistics/FrameTypeData.h
YUViewLib/src/statistics/StatisticsDataPainting.cpp
YUViewLib/src/statistics/StatisticsFileVTMBMS.cpp
YUViewLib/src/statistics/StatisticsType.cpp
YUViewLib/src/statistics/StatisticsType.h
```
Mostly it requires changing QString to std::string. One implication is that the displayed precision in the zoom box then has 6 decimal places. There are ways around it of course:

- Use intermediate ostream. it has set precision operators
- Use printf-variant
- with C++20 there is std::format. 
I don't like the first two. Any issues with using C++20?

I don't think it makes sense to track the changes here. I can make a new branch for it.

ChristianFeldmann commented:
Yep the right way to do this is to open a new Issue for this where we can track all changes that we make regarding the "Qt-freeness" of some parts of the code. I did not have any active work on this planned. Just whenever I touch new code I try to make it Qt free and maybe remove some existing dependencies. 

But maybe we should discuss this a bit before making any rash changes. Like a more planned approach on how to refactor what. I don't think we can got to C++20 yet because the native compilers of Ubuntu 18 does not support it.

nolyn commented:
> Yep the right way to do this is to open a new Issue for this where we can track all changes that we make regarding the "Qt-freeness" of some parts of the code. I did not have any active work on this planned. Just whenever I touch new code I try to make it Qt free and maybe remove some existing dependencies.
> 
> But maybe we should discuss this a bit before making any rash changes. Like a more planned approach on how to refactor what. I don't think we can got to C++20 yet because the native compilers of Ubuntu 18 does not support it.

Ok, I'll just leave that then for now. Doesn't make any sense unless we define first what functionality we want to expose.

nolyn commented:
Seems to work fine now functionality-wise but there are still some stability issues I'm looking into. Crashes when having statistics and zoom box and changing frames rapidly

nolyn committed:
passing by reference made code unstable, passing by value instead

nolyn committed:
added missing lock using statistic datas accessMutex

nolyn committed:
added another lock with accesmutex used when zoombox accesses
statistics data. bringing back references instead of pass by
values. references were not the actual problem (which would have been
strange indeed)

nolyn commented on YUViewLib/src/statistics/StatisticsFileVTMBMS.cpp:
This does not feel right yet. @ChristianFeldmann what do you think? setFrameIndex a line above also uses the mutex and the other uses of the mutex are all handled by the class itself. Thats why it seems strange to put it like this. I think the csv statistic parser should have the same problem. we should check.

nolyn replied:
First time using the review functionality, did it wrong somehow xD

nolyn reviewed (commented):
oh btw. that was an older problem, crashes also happens with the current yuview release

nolyn committed:
added tests for vtmbms statistics: affine MV, GeoPartitioning line, Geo MV (polygon vector)

nolyn commented:
Added missing test. Anything else left before we can merge it?

ChristianFeldmann committed:
Merge branch 'develop' into bugfix/improveVTMstatisticsParser

ChristianFeldmann committed:
Tiny bit of refactoring

ChristianFeldmann committed:
Compile fix
